fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/PersistentStorage.java

```diff
@@ -0,0 +1,78 @@
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * DuraSpace licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.fcrepo.kernel.impl;
+
+import java.time.Instant;
+
+import org.fcrepo.kernel.api.models.FedoraResource;
+import org.fcrepo.kernel.api.models.Tombstone;
+
+/**
+ * Interface for interacting with a persistent storage implementation.
+ *
+ * @author whikloj
+ * @since 2019-09-19
+ */
+public interface PersistentStorage {
+
+    /**
+     * Create a new resource on the persistent storage.
+     *
+     * @param psTxId The persistent storage transaction for this action.
+     * @param resource The new fedora resource to persist to storage.
+     * @return the resource with any new information related to the persisting..?
+     */
+    public FedoraResource create(final PersistentStorageTransaction psTxId, final FedoraResource resource);
+
+    /**
+     * Update an existing resource on persistent storage.
+     *
+     * @param psTxId The persistent storage transaction for this action.
+     * @param resource The changed fedora resource to persist to storage.
+     * @return the updated resource with any new information related to the persisting..?
+     */
+    public FedoraResource update(final PersistentStorageTransaction psTxId, final FedoraResource resource);
+
+    /**
+     * Delete a resource from persistent storage.
+     *
+     * @param psTxId The persistent storage transaction for this action.
+     * @param resource The current fedora resource to delete.
+     * @return The tombstone for the removed resource.
+     */
+    public Tombstone delete(final PersistentStorageTransaction psTxId, final FedoraResource resource);
+
+    /**
+     * Return a resource from persistent storage
+     *
+     * @param psTxId The persistent storage transaction for this action.
+     * @param identifier The identifier of the resource to retrieve.
+     * @return The resource.
+     */
+    public FedoraResource read(final PersistentStorageTransaction psTxId, final String identifier);
+
+    /**
+     * Return a version of a resource from persistent storage
+     *
+     * @param identifier The identifier of the resource to retrieve.
+     * @param version The date/time of the version to retrieve.
+     * @return The version of the resource.
+     */
+    public FedoraResource read(final String identifier, final Instant version);
```

Comment:
I'll note that this does **not** allow for a transaction. I am unsure if it should, because we have said all actions would happen in some sort of transaction.

Reply:
i think it should, since resources created in a transaction are only visible within that transaction

Reply:
right, re: "only visible within that transaction".  It does raise the question as whether an implicit transaction is needed for a basic read, so maybe it can be considered nullable?  or optional?  or a different method?  Not sure.

Reply:
I am okay with `null`ing this for non-transaction requests. That does mean that somewhere higher up we are _not_ creating a transaction...which was not the plan so far. I do wonder if there will be a cumulative performance benefit to not creating transactions unless required.

Reply:
Also this is the `read` for a version of the resource. Versus the `read` further up which would probably be used when accessing a changed resource in a transaction.

However I guess if you are performing a **restore** you might link the `GET` of the Memento and the `PUT` of it over-top the HEAD of the resource together in a transaction.

In which case you would only need to be aware you were _inside_ a transaction if the version you want was created in the same transaction.

Reply:
Weren't we saying the creating a version within a transaction would not be allowed?  It seems to me that when reading committed versions,  a transaction is never needed.  I think that should be explicit in the api.

Reply:
I think different methods for read a version versus reading the current state of the object within an uncommitted transaction is most clear (as it is currently defined)

Reply:
Whether to support creating a version in a transaction may depend on whether or not ending a transaction creates a new version automatically. And that is likely going to be persistence impl specific.

Sorry, I missed that `version` param. I agree, the transaction should generally not matter when getting a version. The only edge case that comes to mind:
* the client deletes an object in a transaction, and then requests a past version of that object while passing along the transaction id (since it would probably be passing in the transaction id on every request). Should it get a 404/410, or should it get the past version?

I'm guessing that in modeshape you could create versions in a transaction, but I don't know if anyone really did that.